File: examples/typescript/case07.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
  ResultHandoff,
  downstreamWithoutCancellation,
  longWorkCheckCancel,
  precheckThenBlock,
} from "./case07.ts";

test("cancel must participate at the handoff, and late cancel is not rollback", async () => {
  const weakCancel = new AbortController();
  const weakHandoff = new ResultHandoff();
  const weak = precheckThenBlock(weakCancel.signal, weakHandoff, 21);
  assert.equal(weakHandoff.hasPending(), true);
  weakCancel.abort();
  assert.equal(weakHandoff.hasPending(), true);
  assert.equal(await weakHandoff.receive(), 42); // Cleanup the intentionally blocked send.
  await weak;

  const strongCancel = new AbortController();
  const strongHandoff = new ResultHandoff();
  const strong = strongHandoff.sendWithCancel(42, strongCancel.signal);
  assert.equal(strongHandoff.hasPending(), true);
  strongCancel.abort();
  await assert.rejects(strong, /CANCELLED/);
  assert.equal(strongHandoff.hasPending(), false);

  const readyCancel = new AbortController();
  const readyHandoff = new ResultHandoff();
  const receiving = readyHandoff.receive();
  await readyHandoff.sendWithCancel(42, readyCancel.signal);
  readyCancel.abort();
  assert.equal(await receiving, 42);

  assert.throws(() => longWorkCheckCancel(strongCancel.signal, 100000), /CANCELLED/);
  let release!: () => void;
  const pending = new Promise<void>((resolve) => { release = resolve; });
  let finished = false;
  const downstream = downstreamWithoutCancellation(pending).then(() => { finished = true; });
  assert.equal(finished, false);
  release();
  await downstream;
  assert.equal(finished, true);
});

File: examples/go/case07_test.go
package examplesgo

import (
	"context"
	"testing"
	"time"
)

func case07Await(t *testing.T, done <-chan error) error {
	t.Helper()
	select {
	case err := <-done:
		return err
	case <-time.After(time.Second):
		t.Fatal("worker failed to terminate")
		return nil
	}
}

func TestCase07CancellationAtBlockingHandoff(t *testing.T) {
	weakCtx, weakCancel := context.WithCancel(context.Background())
	weakOut, weakEntered, weakDone := make(chan int), make(chan struct{}), make(chan error, 1)
	go func() { weakDone <- PrecheckThenBlock(weakCtx, weakOut, 21, weakEntered) }()
	<-weakEntered
	weakCancel()
	select {
	case <-weakDone:
		t.Fatal("weak send should remain blocked without a receiver")
	case <-time.After(10 * time.Millisecond):
	}
	if got := <-weakOut; got != 42 {
		t.Fatalf("cleanup receiver got %d", got)
	}
	if err := case07Await(t, weakDone); err != nil {
		t.Fatal(err)
	}

	strongCtx, strongCancel := context.WithCancel(context.Background())
	strongOut, strongEntered, strongDone := make(chan int), make(chan struct{}), make(chan error, 1)
	go func() { strongDone <- BlockingSendWithCancel(strongCtx, strongOut, 42, strongEntered) }()
	<-strongEntered
	strongCancel()
	if err := case07Await(t, strongDone); err != context.Canceled {
		t.Fatalf("want cancellation, got %v", err)
	}

	readyCtx, readyCancel := context.WithCancel(context.Background())
	readyOut, readyEntered, readyDone := make(chan int, 1), make(chan struct{}), make(chan error, 1)
	go func() { readyDone <- BlockingSendWithCancel(readyCtx, readyOut, 42, readyEntered) }()
	<-readyEntered
	if err := case07Await(t, readyDone); err != nil {
		t.Fatal(err)
	}
	readyCancel() // Late cancellation does not retract a completed result.
	if got := <-readyOut; got != 42 {
		t.Fatalf("late cancel changed delivered result: %d", got)
	}

	if err := LongWorkCheckCancel(strongCtx, 100000); err != context.Canceled {
		t.Fatal("long work ignored cancellation")
	}
	release, entered, downstreamDone := make(chan struct{}), make(chan struct{}), make(chan struct{})
	go func() { DownstreamWithoutCancellation(release, entered); close(downstreamDone) }()
	<-entered
	select {
	case <-downstreamDone:
		t.Fatal("downstream should not finish without release")
	default:
	}
	close(release)
	select {
	case <-downstreamDone:
	case <-time.After(time.Second):
		t.Fatal("downstream cleanup failed")
	}
}
